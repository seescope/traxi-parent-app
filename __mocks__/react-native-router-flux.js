import React, { PropTypes } from 'react';
import { View } from 'react-native';

export const Actions = {
  createKid: jest.fn(),
  walkthrough: jest.fn(),
  congratulations: jest.fn(),
  thankyou: jest.fn(),
  areYouReady: jest.fn(),
  intro: jest.fn(),
};

export const Scene = (props) => <View {...props} />;

export const Router = ({ children }) => (
  <View>{children}</View>
);

Router.propTypes = {
  children: PropTypes.array.isRequired,
};
