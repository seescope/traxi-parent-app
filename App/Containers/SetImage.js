import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Dimensions, Text, View, Alert } from 'react-native';
import ImagePicker from 'react-native-image-picker';
import { Actions } from 'react-native-router-flux';

import { selectKidImage, NEXT_STEP } from '../Actions/Actions';
import setupKid from '../Actions/SetupKid';
import watchDevice from '../Actions/WatchDevice';
import Button from '../Components/Button';
import HeaderText from '../Components/HeaderText';
import Spacing from '../Components/Spacing';
import KidAvatar from '../Components/KidAvatar';
import { WHITE, TRANSPARENT } from '../Constants/Colours';
import { isIOS, logError, firstName } from '../Utils';

const { width } = Dimensions.get('window');

const getSource = response => {
  let source;

  if (response.data) {
    if (isIOS) {
      source = { uri: response.uri.replace('file://', ''), isStatic: true };
    } else {
      source = { uri: response.uri, isStatic: true };
    }
  }

  return source;
};

export const selectImage = pickImage => dispatch => {
  const options = {
    title: 'Select Image',
    storageOptions: {
      skipBackup: true,
    },
  };

  if (!pickImage) {
    dispatch(selectKidImage('http://i.imgur.com/ZrwsRFD.png'));

    return dispatch(setupKid()).then(() => {
      dispatch(NEXT_STEP);
      dispatch(watchDevice());
    });
  }

  return ImagePicker.launchImageLibrary(options, response => {
    if (response.didCancel) {
      return;
    }

    if (response.error) {
      if (response.error === 'Photo library permissions not granted') {
        Alert.alert(
          'Unable to access your photos',
          'Please allow traxi to access your photos to continue.'
        );
      }

      logError(response.error);
      return;
    }

    const source = getSource(response);
    dispatch(selectKidImage(source.uri));
    dispatch(NEXT_STEP);

    dispatch(setupKid()).then(() => {
      dispatch(watchDevice());
    });
  });
};

const style = {
  container: {
    alignItems: 'center',
  },
  bodyText: {
    fontFamily: 'Raleway-Regular',
    fontSize: 16,
    color: WHITE,
    backgroundColor: TRANSPARENT,
    textAlign: 'left',
  },
  buttonContainer: {
    width: width - 64,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
};

const SetImage = ({ parentName, kidName, onPress }) => (
  <View style={style.container}>
    <HeaderText>Thanks, {parentName}!</HeaderText>

    <Spacing height={32} />

    <KidAvatar
      size={204}
      avatarURL=""
    />

    <Spacing height={32} />

    <Text style={style.bodyText}>
      Now, let's set a picture for {kidName}.
    </Text>

    <Spacing />

    <Text style={style.bodyText}>
      Don't worry, only you will be able to see it.
    </Text>

    <Spacing height={32} />

    <View style={style.buttonContainer}>
      <Button primary={false} onPress={() => onPress(false)}>Not right now</Button>
      <Button onPress={() => onPress(true)}>Set a picture for {kidName}</Button>
    </View>

  </View>
);

SetImage.propTypes = {
  kidName: PropTypes.string.isRequired,
  parentName: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  kidName: firstName(state.selectedKid.name),
  parentName: state.parentName,
});

const mapDispatchToProps = {
  onPress: selectImage,
};

export default connect(mapStateToProps, mapDispatchToProps)(SetImage);
