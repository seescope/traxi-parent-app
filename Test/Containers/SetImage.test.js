import React from 'react';
import renderer from 'react-test-renderer';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ImagePicker from 'react-native-image-picker';
import { Actions } from 'react-native-router-flux';

import SetImage, { selectImage } from '../../App/Containers/SetImage';

const TEST_KID = {
  UUID: 'Test UUID',
  name: 'Emanuel Goldstein',
};
const TEST_KIDS = [ TEST_KID ];
const mockStore = configureStore([thunk]);
const testStoreWithoutDeeplink = mockStore({
  parentName: 'Name',
  selectedKid: TEST_KID,
  deeplink: false,
  kids: TEST_KIDS,
  profile: {
    UUID: 'Parent-UUID',
  },
});

const testStoreWithDeeplink = mockStore({
  parentName: 'Name',
  selectedKid: TEST_KID,
  deeplink: true,
  kids: TEST_KIDS,
  profile: {
    UUID: 'Parent-UUID',
  },
});

const SetImageComponentWithDeeplink = () => (
  <Provider store={testStoreWithDeeplink}>
    <SetImage />
  </Provider>
);

const SetImageComponentWithoutDeeplink = () => (
  <Provider store={testStoreWithoutDeeplink}>
    <SetImage />
  </Provider>
);

it('renders the <SetImage> component correctly without deeplink', () => {
  const tree = renderer.create(<SetImageComponentWithoutDeeplink />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the <SetImage> component correctly with deeplink', () => {
  const tree = renderer.create(<SetImageComponentWithDeeplink />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('calls ImagePicker if pickImage is true', () => {
  const dispatch = jest.fn(() => Promise.resolve());
  selectImage(true, true, TEST_KID, TEST_KIDS)(dispatch);

  // Run our assertions on the next tick. Probably more elegant ways of doing this.
  return Promise.resolve().then(() => {
    expect(ImagePicker.launchImageLibrary).toHaveBeenCalled();
    expect(Actions.setupCompletion).toHaveBeenCalled();
  });
});

it('uses a default image of pickImage is false', () => {
  const dispatch = jest.fn(() => Promise.resolve());
  ImagePicker.launchImageLibrary.mockClear();
  selectImage(false)(dispatch);

  // Run our assertions on the next tick. Probably more elegant ways of doing this.
  return Promise.resolve().then(() => {
    expect(ImagePicker.launchImageLibrary).not.toHaveBeenCalled();
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
});
