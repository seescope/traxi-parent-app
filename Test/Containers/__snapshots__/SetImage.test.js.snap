exports[`test calls ImagePicker if pickImage is true 1`] = `
Array [
  Array [
    Object {
      "avatarURL": "heythere",
      "type": "SELECT_KID_IMAGE",
    },
  ],
  Array [
    Object {
      "type": "NEXT_STEP",
    },
  ],
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;

exports[`test renders the <SetImage> component 1`] = `
<View
  style={Object {}}>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "#FFF",
          "fontFamily": "Raleway-ExtraBold",
          "fontSize": 30,
          "marginBottom": -2,
          "textAlign": "center",
        }
      }>
      [missing \"en.general.thanks\" translation]
      , 
      Name
      !
    </Text>
    <View
      style={
        Object {
          "height": 32,
        }
      } />
    <Image
      animation={undefined}
      duration={undefined}
      iterationCount={undefined}
      source={1}
      style={
        Object {
          "borderColor": undefined,
          "borderRadius": 102,
          "borderWidth": 5.828571428571428,
          "height": 204,
          "marginBottom": 20.4,
          "shadowColor": "rgba(0,0,0,0.5)",
          "shadowOffset": Object {
            "height": 2,
          },
          "shadowRadius": 5,
          "width": 204,
        }
      } />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "color": "#FFF",
          "fontFamily": "Raleway-Regular",
          "fontSize": 16,
          "textAlign": "center",
        }
      }>
      [missing \"en.setImage.setAPictureFor\" translation]
       
      Jeff
      .
    </Text>
    <View
      style={
        Object {
          "height": 32,
        }
      } />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "color": "#FFF",
          "fontFamily": "Raleway-Regular",
          "fontSize": 16,
          "textAlign": "center",
        }
      }>
      [missing \"en.setImage.dontWorry\" translation]
      .
    </Text>
    <View
      style={
        Object {
          "height": 24,
        }
      } />
    <View
      style={
        Object {
          "height": 32,
        }
      } />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 686,
      }
    }>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID={undefined}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFF",
            "fontFamily": "Raleway-Regular",
            "fontSize": 20,
          }
        }>
        [missing \"en.setImage.notNow\" translation]
         
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID={undefined}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFF",
            "fontFamily": "Raleway-Regular",
            "fontSize": 20,
          }
        }>
        [missing \"en.setImage.chooseAPicture\" translation]
         
      </Text>
    </View>
  </View>
</View>
`;

exports[`test uses a default image of pickImage is false 1`] = `
Array [
  Array [
    Object {
      "avatarURL": "http://i.imgur.com/ZrwsRFD.png",
      "type": "SELECT_KID_IMAGE",
    },
  ],
  Array [
    Object {
      "type": "NEXT_STEP",
    },
  ],
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;
